pubsub/pubsub.go

```diff
@@ -0,0 +1,208 @@
+// Package pubsub implements a pub-sub model where there is a single publisher
+// (Server) and multiple subscribers.
+//
+// Though you can have multiple publishers by sharing a pointer to a server or
+// by giving the same channel to each publisher and publishing messages from
+// that channel.
+//
+// Clients subscribe for messages, which could be of any type, using a query. A
+// query can be anything as long as it conforms to the interface. See query
+// package.
+//
+// Server will be non-blocking for publishing messages if you provide an
+// OverflowStrategyDrop option. Note sending messages to subscribers is always
+// non-blocking.
+package pubsub
+
+import (
+	"errors"
+
+	cmn "github.com/tendermint/tmlibs/common"
+)
+
+type operation int
+
+const (
+	sub operation = iota
+	pub
+	unsub
+	shutdown
+)
+
+type overflowStrategy int
+
+const (
+	drop overflowStrategy = iota
+	wait
+)
+
+var (
+	ErrorOverflow = errors.New("Server overflowed")
+)
+
+type cmd struct {
+	op    operation
+	query Query
+	ch    chan<- interface{}
+	msg   interface{}
+	tags  map[string]interface{}
+}
+
+// Query defines an interface for a query to be used for subscribing.
+type Query interface {
+	Matches(tags map[string]interface{}) bool
+}
+
+// Server allows clients to subscribe/unsubscribe for messages, pubsling
+// messages with or without tags, and manages internal state.
+type Server struct {
+	cmn.BaseService
+
+	cmds             chan cmd
+	overflowStrategy overflowStrategy
+}
+
+// NewServer returns a new server. You can provide additional options such as
+// BufferCapacity(N) or OverflowStrategy{X}. See their documentation for
+// details.
+func NewServer(options ...func(*Server)) *Server {
+	s := &Server{overflowStrategy: wait}
+	s.BaseService = *cmn.NewBaseService(nil, "PubSub", s)
+
+	for _, option := range options {
+		option(s)
+	}
+
+	if s.cmds == nil { // if BufferCapacity was not set, create unbuffered channel
+		s.cmds = make(chan cmd)
+	}
+
+	return s
+}
+
+// BufferCapacity allows you to specify capacity for the internal servers
+// queue. Since the server given Y subscribers could only process X messages,
+// this option could be used to survive spikes (e.g. high amount of
+// transactions during peak hours).
```

Comment:
We could run out of memory if there are too many subscribers. Otherwise, according to the Little's law https://en.wikipedia.org/wiki/Little%27s_law and queuing theory

suppose 100 subscribers
λ = 1000 events per second
W = 10ms to match event tags against all 100 subscribers

then L = 1000events/sec * 10ms = 1events/ms * 10 = 10

i.e. an average number of events in the server's queue - 10 and we need BuffersCapacity set to 10 or better 20 to accommodate spikes (like when the arrival rate is 9000 events per second, not usual 1000)

Hope that helps. Or maybe I am wrong here? Anyway, I don't think we will hit the place where we will start dropping events (especially, if we'll optimize it), but every time I design something, I think about that kind of stuff and try to design it the way so nothing super bad happens.